src/main/java/fi/vm/yti/datamodel/api/v2/service/SchemaService.java

```diff
@@ -1,161 +1,264 @@
+	 * @param schemaPID The schema PID.
+	 * @param model     The RDF model.
+	 */
 	private void handleObject(String propID, JsonNode node, String schemaPID, Model model) {
 
-		Resource nodeShapeResource = model.createResource(schemaPID + "#" + propID);
+		String propIDCapitalised = capitaliseNodeIdentifier(propID);
+		String nameProperty = propID.substring(propID.lastIndexOf("/") + 1);
+		Resource nodeShapeResource = model.createResource(schemaPID + "#" + propIDCapitalised);
 		
-		nodeShapeResource.addProperty(RDF.type, model.getResource(SHACL.NodeShape.getURI()));
-		nodeShapeResource.addProperty(MSCR.localName, propID);
-		if(node.has("description"))
+		nodeShapeResource.addProperty(RDF.type, model.getResource(SH.NodeShape.getURI()));
+		nodeShapeResource.addProperty(MSCR.localName, nameProperty);
+		nodeShapeResource.addProperty(model.getProperty(SH.name.getURI()), nameProperty);
+		
+		if (node.has("description"))
 			nodeShapeResource.addProperty(DCTerms.description, node.get("description").asText());
-		nodeShapeResource.addProperty(model.getProperty(SHACL.name.getURI()), propID);
-		if(node == null || node.get("properties") == null) {
+		if (node == null || node.get("properties") == null) 
 			return;
-		}
-		
-		/* Iterate over properties
-		 * If a property is an array or object – add and recursively iterate over them
-		 * If a property is a datatype / literal – it's just added.
+		if (node.has("additionalProperties"))
+			nodeShapeResource.addProperty(SH.closed, model.createTypedLiteral(!node.get("additionalProperties").asBoolean()));
+		/*
+		 * Iterate over properties If a property is an array or object – add and
+		 * recursively iterate over them. If a property is a datatype or literal – it's just added.
 		 */
+
 		Iterator<Entry<String, JsonNode>> propertiesIterator = node.get("properties").fields();
-		while(propertiesIterator.hasNext()) {
+		while (propertiesIterator.hasNext()) {
 			Entry<String, JsonNode> entry = propertiesIterator.next();
-			if(entry.getKey().startsWith("_") || entry.getKey().startsWith("$"))
+			
+			if (entry.getKey().startsWith("_") || entry.getKey().startsWith("$"))
 				continue;
-			//System.out.println(entry.getValue());
-			if(entry.getValue().get("type") != null && entry.getValue().get("type").asText().equals("object")) {
-				// add object property 
-				Resource propertyShape = addObjectProperty(entry.getKey(), entry.getValue(), model, schemaPID, schemaPID + "#" + propID + "/" + entry.getKey());
-				nodeShapeResource.addProperty(model.getProperty(SHACL.property.getURI()), propertyShape);
-				handleObject(propID + "/" + entry.getKey(), entry.getValue(), schemaPID, model);
-			}
-			else if(entry.getValue().get("type") != null && entry.getValue().get("type").asText().equals("array") && entry.getValue().get("items").has("type") && entry.getValue().get("items").get("type").asText().equals("object")) {
+			if (entry.getValue().get("type") == null) 
```

Comment:
Is type a required property in the metadata schema? If not, there should be a default type (string).